client/src/sync-filesystem.js

```diff
@@ -6,39 +6,216 @@
 
 var Filer = require('../../lib/filer.js');
 var Shell = require('../../lib/filer-shell.js');
-var Path = Filer.Path;
 var fsUtils = require('../../lib/fs-utils.js');
 var conflict = require('../../lib/conflict.js');
-var resolvePath = require('../../lib/sync-path-resolver.js').resolve;
+var syncModes = require('../../lib/constants.js').syncModes;
+var findPathIndexinArray = require('../../lib/util.js').findPathIndexinArray;
 
 function SyncFileSystem(fs) {
   var self = this;
-  var pathToSync;
-  var modifiedPath;
-
-  // Manage path resolution for sync path
-  Object.defineProperty(self, 'pathToSync', {
-    get: function() { return pathToSync; },
-    set: function(path) {
-      if(path) {
-        pathToSync = resolvePath(pathToSync, path);
-      } else {
-        pathToSync = null;
-      }
+  var root = '/';
+  // Record changes during a downstream sync
+  // Is a parallel array with sourceListCache
+  var recordedPaths = [];
+  var sourceListCache = [];
+  var changesDuringDownstream = [];
+
+  // Expose the root used to sync for the filesystem
+  // Defaults to '/'
+  Object.defineProperties(self, {
+    'root': {
+      get: function() { return root; }
+    },
+    'changesDuringDownstream': {
+      get: function() { return changesDuringDownstream; }
     }
   });
 
-  // Record modifications to the filesystem during a sync
-  Object.defineProperty(fs, 'modifiedPath', {
-    get: function() { return modifiedPath; },
-    set: function(path) {
-      if(path) {
-        modifiedPath = resolvePath(modifiedPath, path);
-      } else {
-        modifiedPath = null;
+  self.record = function(path, sourceList) {
+    recordedPaths.push(path);
+    sourceListCache.push(sourceList);
+  };
+
+  self.stopRecording = function(path) {
+    var indexInRecordedPaths = recordedPaths.indexOf(path);
+    recordedPaths.splice(indexInRecordedPaths, 1);
+    changesDuringDownstream.splice(changesDuringDownstream.indexOf(path), 1);
+    return sourceListCache.splice(indexInRecordedPaths, 1)[0];
+  };
+
+  // Get the paths queued up to sync
+  self.getPathsToSync = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
       }
+
+      callback(null, pathsToSync && pathsToSync.toSync);
+    });
+  };
+
+  // Add paths to the sync queue where paths is an array
+  self.appendPathsToSync = function(paths, callback) {
+    if(!paths || !paths.length) {
+      return callback();
     }
-  });
+
+    paths = paths.filter(function(path) {
+      return path.indexOf(root) === 0;
+    });
+
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      var toSync = pathsToSync.toSync;
```

Comment:
Above you did an extra check to make sure `pathsToSync` also existed.  Is that necessary here too?  Will `pathsToSync` ever be null?